stratosphere/ncm/source/ncm_content_meta_database.cpp

```diff
@@ -0,0 +1,584 @@
+        R_TRY(this->EnsureEnabled());
+        
+        size_t entries_total = 0;
+        size_t entries_written = 0;
+
+        /* If there are no entries then we've already successfully listed them all. */
+        if (this->kvs->GetCount() == 0) {
+            out_entries_total.SetValue(entries_total);
+            out_entries_written.SetValue(entries_written);
+            return ResultSuccess();
+        }
+
+        for (auto entry = this->kvs->begin(); entry != this->kvs->end(); entry++) {
+            ContentMetaKey key = entry->GetKey();
+
+            /* Check if this entry matches the given filters. */
+            if (!((static_cast<u8>(type) == 0 || key.type == type))) {
+                continue;
+            }
+
+            const void* value = nullptr;
+            size_t value_size = 0;
+            R_TRY(GetContentMetaValuePointer(&value, &value_size, key, this->kvs));
+
+            if (key.type == ContentMetaType::Application || key.type == ContentMetaType::Patch || key.type == ContentMetaType::AddOnContent || key.type == ContentMetaType::Delta) {
+                ProgramId application_id = key.id;
+                
+                switch (key.type) {
+                    case ContentMetaType::Application:
+                        break;
+                    default:
+                        /* The first u64 of all non-application extended headers is the application program id. */
+                        application_id = *GetValueExtendedHeader<ProgramId>(value);
+                }
+
+                /* Write the entry to the output buffer. */
+                if (entries_written < out_keys.GetSize()) {
+                    ApplicationContentMetaKey* out_app_key = &out_keys[entries_written];
+                    out_app_key->application_program_id = application_id;
+                    out_app_key->key = key;
+                    entries_written++;
+                }
+
+                entries_total++;
+            }
+        }
+
+        out_entries_total.SetValue(entries_total);
+        out_entries_written.SetValue(entries_written);
+        return ResultSuccess();
+    }
+
+    Result ContentMetaDatabaseInterface::Has(sf::Out<bool> out, ContentMetaKey key) {
+        R_TRY(this->EnsureEnabled());
+
+        if (this->kvs->GetCount() == 0) {
+            out.SetValue(false);
+            return ResultSuccess();
+        }
+
+        bool has = false;
+        const auto it = this->kvs->lower_bound(key);
+        if (it != this->kvs->end()) {
+            has = it->GetKey() == key;
+        }
+
+        out.SetValue(has);
+        return ResultSuccess();
+    }
+
+    Result ContentMetaDatabaseInterface::HasAll(sf::Out<bool> out, const sf::InArray<ContentMetaKey> &keys) {
+        R_TRY(this->EnsureEnabled());
+
+        bool has = true;
+        for (size_t i = 0; i < keys.GetSize() && has; i++) {
+            R_TRY(this->Has(&has, keys[i]));
```

Comment:
This function won't work correctly.

It will return whether the last element is present.

What you really mean is like

```cpp
*out = false;

for (size_t i = 0; i < keys.GetSize(); i++) {
    bool has;
    R_TRY(this->Has(std::addressof(has), keys[i]));
    R_UNLESS(has, ResultSuccess());
}

*out = true;
return ResultSuccess();
```